### ISSUE TITLE
Monstache AWS DocumentDB Error

### ISSUE THREAD

**Issue by aspire-shatheesh** [rwynn/monstache#297]
I have tried Monstache in my local development environment and it worked well. But when I try to setup the same in AWS and Document DB it is not working. Could you help me if I should consider changing any other configuration here.

I am getting the following error when I try to connect to AWS document DB. Currently we have turned off the SSL for testing purpose. Could you please help on this issue.

**Observations:**

1. It looks like either during first time or while killing the monstache the data is pushed to elastic successfully, but I could see the below mentioned errors getting emitted continuously.
2. Continuous data sync is not happening, but the said error is flowing in console
2.I tried with resume = true, still there is no change

```
monstache -verbose -f monstache_cfg.toml
WARN 2019/10/22 11:12:17 For performance reasons it is recommended to use elasticsearch-max-bytes instead of elasticsearch-max-docs since doc size may vary
INFO 2019/10/22 11:12:17 Started monstache version 6.2.2
INFO 2019/10/22 11:12:17 MongoDB go driver v1.1.2
INFO 2019/10/22 11:12:17 Elasticsearch go driver 7.0.6
INFO 2019/10/22 11:12:17 Successfully connected to MongoDB version 3.6.0
INFO 2019/10/22 11:12:17 Successfully connected to Elasticsearch version 7.1.1
INFO 2019/10/22 11:12:17 Listening for events
INFO 2019/10/22 11:12:17 Sending systemd READY=1
WARN 2019/10/22 11:12:17 Systemd notification not supported (i.e. NOTIFY_SOCKET is unset)
INFO 2019/10/22 11:12:17 Watching changes on the deployment
INFO 2019/10/22 11:12:17 Direct reads completed
ERROR 2019/10/22 11:12:17 Error starting change stream. Will retry: BSON field '$changeStream.allChangesForCluster' is an unknown field.
ERROR 2019/10/22 11:12:17 Error starting change stream. Will retry: BSON field '$changeStream.allChangesForCluster' is an unknown field.
ERROR 2019/10/22 11:12:17 Error starting change stream. Will retry: BSON field '$changeStream.allChangesForCluster' is an unknown field.
.
.
.
.
.
```

Below is the monstache file
Command used : monstache -verbose -f monstache_cfg.toml

Monstache File:                                                      
```
mongo-url = "mongodb://XXXX.docdb.amazonaws.com:27017"
elasticsearch-urls = ["https://searchXXXXX.amazonaws.com"]
direct-read-namespaces = ["example.example"]
elasticsearch-max-conns = 2
elasticsearch-max-docs = 100
[[mapping]]
namespace = "example.example"
index = "example"
```
`

**Comment by rwynn** [comment]
Hi @aspire-shatheesh,

My understanding is that Amazon DocumentDB supports many but not all MongoDB 3.6 features.  Notably, change streams and gridfs are [not supported](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html).  

Monstache uses the MongoDB change streams API to accomplish real time sync to Elasticsearch.  So, you would not be able to take advantage of this feature with DocumentDB.  MongoDB Atlas (which can deploy clusters to AWS), on the other hand would give you access to these features.  

If you need to use DocumentDB, you may still be able to use monstache to periodically do a full sync to Elasticsearch using `direct-read-namespaces` as you have.  

You can use a config like the one you have currently but add the following:

```toml
disable-change-streams = true
exit-after-direct-reads = true
```

Then with that in place, you would need to run monstache periodically, and it would do a full sync each time.

**Comment by aspire-shatheesh** [comment]
Hi @rwynn 

Thanks for prompt response.  Looks like the key "disable-change-streams" is not available in documentation  too. Am getting the error when i use it.  Did you mean "disable-change-events".

`ERROR 2019/10/22 14:46:10 Config file contains undecoded keys: ["disable-change-streams"]`

Also, Our system will have large data sets about 5 - 10 million. So will it be scalable if we do full sync every time. Does the resume work? If so, I will at least schedule the monstache every minute / 5 mins once.disable-change-events

**Comment by rwynn** [comment]
You are correct, my memory failed, the config is called `disable-change-events`.

The `resume`, `replay` and `cluster-name` options will not work I'm afraid, because they rely on the change streams feature, specifically, the `startAtOperationTime` flag that was added in MongoDB 4.0.

I guess whether or not a full sync every time would work for you depends on how fresh you need the data to be and how quickly your full data set is being synced.

**Comment by aspire-shatheesh** [comment]
Okay. Thanks a lot for the suggestions and helping around @rwynn . I will explore further on if anything can be done with DocumentDB w.r.t syncing large data sets.

**Comment by aspire-shatheesh** [comment]
Hi @rwynn 

We have set up a Mongo V 4.x in ec2 box instead of document DB to use it with monstache and the tool is super awesome, a big thanks to you for it. One day I will learn golang and contribute it back for sure.

Looks like AWS has recently started supporting Change Streams.

https://aws.amazon.com/about-aws/whats-new/2019/10/amazon-documentdb-with-mongodb-compatibility-adds-support-for-change-streams/
https://docs.aws.amazon.com/documentdb/latest/developerguide/change-streams.html

But like you said, the streams API of document DB doesn't have startAtOperationTime yet because I believe the change stream version of AWS DB is 3.6.

`ERROR 2019/11/16 17:02:49 Error starting change stream. Will retry: BSON field '$changeStream.startAtOperationTime' is an unknown field.`

The question I wanted to get it  clarified with you is, is there any way I can do real-time sync with AWS document DB using monstache now since it started supporting change streams.

**Comment by rwynn** [comment]
Hi @aspire-shatheesh thanks for the update, that is pretty cool.  With that update to DocumentDB you should now be able to use the following for real time sync of collections.

```toml
# for 3.6 compatibility you need namespaces of collections here, not db or cluster
change-stream-namespaces = [ "db1.col1", "db2.col2" ]
```
As you mentioned, monstache is currently setup to use timestamps for its `resume` and `cluster-name` capabilities.  So, you would not be able to use those features ... yet.

I have an idea for a new resuming mode that might be compatible with the resuming supported in 3.6.  I just did some tests and it's looking promising.  When I get some time I will release a new version of monstache that supports 2 resume strategies.  Then you should be able to use it hopefully with DocumentDB enabling the `resume` and `cluster-name` features to work.

The only change you will need to make is...

```toml
# resume using tokens instead of timestamps to support 3.6 API
resume-strategy = 1
```

**Comment by aspire-shatheesh** [comment]
That would be great addition @rwynn.  I really appreciate your time. Thanks again

**Comment by rwynn** [comment]
@aspire-shatheesh you can try resume for Amazon DocumentDB now with the latest monstache releases - `v5.4.0` and `v6.4.0`.  

See the [docs](https://rwynn.github.io/monstache-site/advanced/#amazon-documentdb-with-mongodb-compatibility) for instructions.

**Comment by rwynn** [comment]
@aspire-shatheesh let me know if you get a chance to try latest monstache against DocumentDB.  Interested to see how it works with `resume-strategy` = 1.

**Comment by zhengdai** [comment]
> Hi @rwynn
> 
> We have set up a Mongo V 4.x in ec2 box instead of document DB to use it with monstache and the tool is super awesome, a big thanks to you for it. One day I will learn golang and contribute it back for sure.
> 
> Looks like AWS has recently started supporting Change Streams.
> 
> https://aws.amazon.com/about-aws/whats-new/2019/10/amazon-documentdb-with-mongodb-compatibility-adds-support-for-change-streams/
> https://docs.aws.amazon.com/documentdb/latest/developerguide/change-streams.html
> 
> But like you said, the streams API of document DB doesn't have startAtOperationTime yet because I believe the change stream version of AWS DB is 3.6.
> 
> `ERROR 2019/11/16 17:02:49 Error starting change stream. Will retry: BSON field '$changeStream.startAtOperationTime' is an unknown field.`
> 
> The question I wanted to get it clarified with you is, is there any way I can do real-time sync with AWS document DB using monstache now since it started supporting change streams.

Hello, can you tell me does the documentdb support monstache after you try？

**Comment by aspire-shatheesh** [comment]
Sorry @rwynn  somehow got lost in the thread of emails. We were apprehensive to use the aws document DB considering the rich amount of aggregations that we would do. Also we did want to get handicapped by AWS Document DB not being able to support the latest functions. So we went with mongo box on ec2. Closing this thread since we are not using it.
But thanks again for support

**Comment by zhengdai** [comment]
> Sorry @rwynn somehow got lost in the thread of emails. We were apprehensive to use the aws document DB considering the rich amount of aggregations that we would do. Also we did want to get handicapped by AWS Document DB not being able to support the latest functions. So we went with mongo box on ec2. Closing this thread since we are not using it.
> But thanks again for support

I tried it on documentdb and it works well. what's your handicap?

**Comment by aspire-shatheesh** [comment]
@zhengdai  when I  wrote that initially about Document DB it was 1 year back and they didn't had support for change streams and in the mid of this thread they introduced change streams so the document DB that we see now is much better than what it was earlier an year.  When I say handicapped, still they are officially supporting only V3.6 of mongo with some features that is available in Official mongo v4.2 and [they dont support many mongo](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html) stage aggregation functions such as $out, $replaceRoot, $$ROOT and etc which we use extensively

**Comment by zhengdai** [comment]
> @zhengdai when I wrote that initially about Document DB it was 1 year back and they didn't had support for change streams and in the mid of this thread they introduced change streams so the document DB that we see now is much better than what it was earlier an year. When I say handicapped, still they are officially supporting only V3.6 of mongo with some features that is available in Official mongo v4.2 and [they dont support many mongo](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html) stage aggregation functions such as $out, $replaceRoot, $$ROOT and etc which we use extensively

You can consider buy aliyun mongodb, it's now 4.2, hahaha

**Comment by pcinnusamy** [comment]
### 
**@rwynn @aspire-shatheesh @zhengdai ,

Your helping hands would be very welcome to remove the barrier at this moment :

 We have done the DocumentDB3.6 setup on our AWS cluster and created end points with replica set rs0 on primary .
Also , we have created the separate cluster for Monstache service with EC2 instance and performed connectivity testing between DocumentDB cluster and ES cluster, but we are unable to connect while running thru Monstahce configuration. PFA results for your reference.

Tring to connect DocumentDB from Monstahce cluster via telnet  :(connection between db is open)
========================================================
]# telnet mongodb-poc-cluster-***************.amazonaws.com 27017
Trying ipaddress.
Connected to mongodb-poc-cluster-************.docdb.amazonaws.com.
Escape character is '^]'.

Steps followed for Monstache installation :
==============================
1.wget https://dl.google.com/go/go1.14.4.linux-amd64.tar.gz 
2. tar -C /usr/local -xzf go1.14.4.linux-amd64.tar.gz
3. Configure environment variables.
Run the vim /etc/profile command to open the configuration file for environment variables. Then, add the following content to the file. GOPROXY specifies a proxy for the modules of Alibaba Cloud SDK for Go.
export GOROOT=/usr/local/go
export GOPATH=/home/go/
export PATH=$PATH:$GOROOT/bin:$GOPATH/bin
4.Apply the environment variables.
source /etc/profile
Go to the installation path.
cd /usr/local/
5. Download the installation package from the GitHub repository.
git clone https://github.com/rwynn/monstache.git 
6. Go to the monstache directory.
cd monstache
7. git checkout rel6
8.Install Monstache.
go install
9.  monstache -v
6.7.4
10. cd /usr/local/monstache/
vim config.toml
11. Modify the configuration file.
PFA file.
12.After placed configuration file , Run Monstache.
monstache -f config.toml

 Executing  Monstache service for accessing DocumentDB :
 =========================
[root@ip- /doc/poc/monstache/monstache]# monstache -f config.toml
ERROR 2021/02/18 12:07:49 Unable to connect to MongoDB using URL mongodb://R**EDACTED@mongodb-poc-cluster.cluster-***********.amazonaws.com:27017/?authSource=admin&replicaSet=rs0&ssl_ca_certs=rds-combined-ca-bundle.pem: server selection error: server selection timeout, current topology: { Type: ReplicaSetNoPrimary, Servers: [{ Addr: mongodb-poc-cluster.cluster-chxlzpkqm5yc.us-east-1.docdb.amazonaws.com:27017, Type: Unknown, State: Connected, Average RTT: 0, Last error: connection() : connection(mongodb-poc-cluster.cluster-**********.amazonaws.com:27017[-5]) incomplete read of message header: read tcp 10.31.35.125:39722->10.31.9.164:27017: i/o timeout }, ] }**

**Executing Monstache service for accessing MongoDB :( working without any issue )**
========================================================
Configuration : 
==========

config.toml

mongo-url="mongodb://mongopoc_admin:********@mongodb-poc-cluster.cluster-*************.amazonaws.com:27017/?authSource=admin&replicaSet=rs0&ssl_ca_certs=rds-combined-ca-bundle.pem"
elasticsearch-urls = ["https://********************:9200"]
direct-read-namespaces = ["admin.Docdbtoes"]
change-stream-namespaces = ["admin.Docdbtoes"]
gzip = true
stats = true
index-stats = true
elasticsearch-user = "poc-documentdb"
elasticsearch-password = "******************"
elasticsearch-max-conns = 4
elasticsearch-pem-file = "/doc/poc/monstache/monstache/pem/es.pem"
elastic-validate-pem-file = true
dropped-collections = true
dropped-databases = true
resume = true
resume-strategy = 0
index-files = true
exit-after-direct-reads =true**
